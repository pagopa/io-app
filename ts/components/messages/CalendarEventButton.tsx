import { Text } from "native-base";
import React from "react";
import { Alert, Dimensions, StyleSheet } from "react-native";
import { Calendar } from "react-native-calendar-events";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import { CreatedMessageWithContent } from "../../../definitions/backend/CreatedMessageWithContent";
import I18n from "../../i18n";
import {
  addCalendarEvent,
  AddCalendarEventPayload,
  removeCalendarEvent,
  RemoveCalendarEventPayload
} from "../../store/actions/calendarEvents";
import { preferredCalendarSaveSuccess } from "../../store/actions/persistedPreferences";
import {
  CalendarEvent,
  calendarEventByMessageIdSelector
} from "../../store/reducers/entities/calendarEvents/calendarEventsByMessageId";
import { preferredCalendarSelector } from "../../store/reducers/persistedPreferences";
import { GlobalState } from "../../store/reducers/types";
import { openAppSettings } from "../../utils/appSettings";
import {
  checkAndRequestPermission,
  isEventInCalendar,
  removeCalendarEventFromDeviceCalendar,
  saveCalendarEvent,
  searchEventInCalendar
} from "../../utils/calendar";
import ButtonDefaultOpacity from "../ButtonDefaultOpacity";
import { withLightModalContext } from "../helpers/withLightModalContext";
import SelectCalendarModal from "../SelectCalendarModal";
import IconFont from "../ui/IconFont";
import { LightModalContextInterface } from "../ui/LightModal";

type OwnProps = {
  message: CreatedMessageWithContent;
  small?: boolean;
  medium?: boolean;
  disabled?: boolean;
};

type Props = OwnProps &
  ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps> &
  LightModalContextInterface;

type State = Readonly<{
  // Store if the event is in the device calendar
  isEventInDeviceCalendar: boolean;
}>;

const screenWidth = Dimensions.get("window").width;
const minScreenWidth = 320;
// On small devices use short label
const reminderText = I18n.t(
  screenWidth <= minScreenWidth
    ? "messages.cta.reminderShort"
    : "messages.cta.reminder"
);

const styles = StyleSheet.create({
  button: {
    flex: 1
  },
  oneThird: {
    flex: 5
  }
});

class CalendarEventButton extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      isEventInDeviceCalendar: false
    };
  }

  /**
   * A function to check if the eventId of the CalendarEvent stored in redux
   * is really/still in the device calendar.
   * It is important to make this check because the event can be removed outside
   * the App.
   */
  private checkIfEventInCalendar = async (
    calendarEvent: CalendarEvent | undefined
  ) => {
    if (calendarEvent === undefined) {
      this.setState({
        isEventInDeviceCalendar: false
      });
      return;
    }
    const mayBeInCalendar = await isEventInCalendar(
      calendarEvent.eventId
    ).run();
    this.setState({
      isEventInDeviceCalendar: mayBeInCalendar.fold(_ => false, s => s)
    });
  };

  public async componentDidMount() {
    const { calendarEvent } = this.props;

    // If we have a calendar event in the store associated to this message
    // Check if the event is still in the device calendar
    await this.checkIfEventInCalendar(calendarEvent);
  }

  public async componentDidUpdate(prevProps: Props) {
    // if calenderEvent changes means reminder has been changed
    if (prevProps.calendarEvent !== this.props.calendarEvent) {
      // if a calendarEvent exists we have to check if it really exists as calendar event
      // the event can be removed outside the App.
      await this.checkIfEventInCalendar(this.props.calendarEvent);
    }
  }

  private confirmSaveCalendarEventAlert = (
    calendar: Calendar,
    message: CreatedMessageWithContent,
    dueDate: Date,
    title: string,
    eventId: string
  ) =>
    Alert.alert(
      I18n.t("messages.cta.reminderAlertTitle"),
      I18n.t("messages.cta.reminderAlertDescription"),
      [
        {
          text: I18n.t("global.buttons.cancel"),
          style: "cancel"
        },
        {
          text: I18n.t("messages.cta.reminderAlertKeep"),
          style: "default",
          onPress: () =>
            this.setState(
              {
                isEventInDeviceCalendar: true
              },
              () => {
                // Add the calendar event to the store
                this.props.addCalendarEvent({
                  messageId: message.id,
                  eventId
                });
              }
            )
        },
        {
          text: I18n.t("messages.cta.reminderAlertAdd"),
          style: "default",
          onPress: () =>
            saveCalendarEvent(
              calendar,
              message,
              dueDate,
              title,
              this.addCalendarEvent
            )
        }
      ],
      { cancelable: false }
    );

  // Add the calendar event to the store
  private addCalendarEvent = (calendarEvent: AddCalendarEventPayload) => {
    this.props.addCalendarEvent({
      messageId: calendarEvent.messageId,
      eventId: calendarEvent.eventId
    });

    this.setState({
      isEventInDeviceCalendar: true
    });
  };

  private addCalendarEventToDeviceCalendar = (
    message: CreatedMessageWithContent,
    dueDate: Date
  ) => (calendar: Calendar) => {
    const title = I18n.t("messages.cta.reminderTitle", {
      title: message.content.subject
    });

    const { preferredCalendar } = this.props;

    this.props.hideModal();

    if (preferredCalendar === undefined) {
      this.props.preferredCalendarSaveSuccess(calendar);
    }

    searchEventInCalendar(dueDate, title)
      .then(mayBeEventId =>
        mayBeEventId.foldL(
          async () => {
            await saveCalendarEvent(
              calendar,
              message,
              dueDate,
              title,
              this.addCalendarEvent
            );
          },
          async eventId => {
            this.confirmSaveCalendarEventAlert(
              calendar,
              message,
              dueDate,
              title,
              eventId
            );
          }
        )
      )
      .catch(() =>
        saveCalendarEvent(calendar, message, dueDate, title, addCalendarEvent)
      );
  };

  private onRemoveCalendarEvent = (calendarEvent: CalendarEvent) => {
    this.props.removeCalendarEvent({
      messageId: calendarEvent.messageId
    });
    this.setState({
      isEventInDeviceCalendar: false
    });
  };

  // Create an action to add or remove the event
  private onPressHandler = () => {
    const { calendarEvent, message, preferredCalendar } = this.props;
    const { due_date } = message.content;

    if (due_date === undefined) {
      return;
    }

    // Check the authorization status
    checkAndRequestPermission()
      .then(calendarPermission => {
        if (calendarPermission.authorized) {
          if (this.state.isEventInDeviceCalendar) {
            // If the event is in the calendar prompt an alert and ask for confirmation
            Alert.alert(
              I18n.t("messages.cta.reminderRemoveRequest.title"),
              undefined,
              [
                {
                  text: I18n.t("messages.cta.reminderRemoveRequest.cancel"),
                  style: "cancel"
                },
                {
                  text: I18n.t("messages.cta.reminderRemoveRequest.ok"),
                  style: "destructive",
                  onPress: () => {
                    // after confirmation remove it
                    removeCalendarEventFromDeviceCalendar(
                      calendarEvent,
                      this.onRemoveCalendarEvent
                    );
                  }
                }
              ],
              { cancelable: false }
            );
          } else if (preferredCalendar !== undefined) {
            this.addCalendarEventToDeviceCalendar(message, due_date)(
              preferredCalendar
            );
          } else {
            // The event need to be added
            // Show a modal to let the user select a calendar
            this.props.showModal(
              <SelectCalendarModal
                onCancel={this.props.hideModal}
                onCalendarSelected={this.addCalendarEventToDeviceCalendar(
                  message,
                  due_date
                )}
              />
            );
          }
        } else if (!calendarPermission.asked) {
          // Authorized is false (denied, restricted or undetermined)
          // If the user denied permission previously (not in this session)
          // prompt an alert to inform that his calendar permissions could have been turned off
          Alert.alert(
            I18n.t("messages.cta.calendarPermDenied.title"),
            undefined,
            [
              {
                text: I18n.t("messages.cta.calendarPermDenied.cancel"),
                style: "cancel"
              },
              {
                text: I18n.t("messages.cta.calendarPermDenied.ok"),
                style: "default",
                onPress: () => {
                  // open app settings to turn on the calendar permissions
                  openAppSettings();
                }
              }
            ],
            { cancelable: true }
          );
        }
      })
      // No permission to add/remove the reminder
      .catch();
  };

  public render() {
    const { small, disabled, medium } = this.props;
    const iconName = this.state.isEventInDeviceCalendar
      ? "io-tick-big"
      : "io-plus";
    return (
      <ButtonDefaultOpacity
        disabled={disabled}
        onPress={this.onPressHandler}
        xsmall={small}
        small={medium}
        bordered={!disabled}
        style={this.props.small ? styles.oneThird : styles.button}
      >
        <IconFont name={iconName} />
        <Text>{reminderText}</Text>
      </ButtonDefaultOpacity>
    );
  }
}

const mapStateToProps = (state: GlobalState, ownProps: OwnProps) => ({
  preferredCalendar: preferredCalendarSelector(state),
  calendarEvent: calendarEventByMessageIdSelector(ownProps.message.id)(state)
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  removeCalendarEvent: (calendarEvent: RemoveCalendarEventPayload) =>
    dispatch(removeCalendarEvent(calendarEvent)),
  addCalendarEvent: (calendarEvent: AddCalendarEventPayload) =>
    dispatch(addCalendarEvent(calendarEvent)),
  preferredCalendarSaveSuccess: (calendar: Calendar) =>
    dispatch(
      preferredCalendarSaveSuccess({
        preferredCalendar: calendar
      })
    )
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withLightModalContext(CalendarEventButton));
