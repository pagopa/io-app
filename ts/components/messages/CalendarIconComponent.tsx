import * as React from "react";
import { G, Path, Svg, Text } from "react-native-svg";

type Props = {
  month: string;
  day: string;
  backgroundColor: string;
  textColor: string;
  small?: boolean;
};

class CalendarIconComponent extends React.PureComponent<Props, never> {
  get dimension(): number {
    return this.props.small ? 32 : 40;
  }

  public render() {
    const { month, day, backgroundColor, textColor } = this.props;
    return (
      <Svg height={this.dimension} width={this.dimension} viewBox="0 0 40 40">
        <G id="calendar">
          <Path
            d="m 33,0 h 3 c 2.209139,0 4,1.790861 4,4 v 32 c 0,2.209139 -1.790861,4 -4,4 H 4 C 1.790861,40 0,38.209139 0,36 V 4 C 0,1.790861 1.790861,0 4,0 h 3 v 3 h 4 V 0 h 18 v 3 h 4 z"
            id="icon"
            fill={backgroundColor}
          />
          <Text
            textAnchor="middle"
            alignmentBaseline="central"
            x="50%"
            y="30%"
            fill={textColor}
            fontSize={10}
          >
            {month}
          </Text>
          <Text
            textAnchor="middle"
            alignmentBaseline="central"
            x="50%"
            y="65%"
            fill={textColor}
            fontSize={18}
            fontWeight="bold"
          >
            {day}
          </Text>
        </G>
      </Svg>
    );
  }
}

export default CalendarIconComponent;
