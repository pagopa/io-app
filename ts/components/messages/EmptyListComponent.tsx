import { Text, View } from "native-base";
import * as React from "react";
import { Image, ImageSourcePropType, StyleSheet } from "react-native";
import customVariables from "../../theme/variables";

const styles = StyleSheet.create({
  view: {
    padding: customVariables.contentPadding,
    alignItems: "center"
  },
  title: {
    paddingTop: customVariables.contentPadding
  },
  subtitle: {
    textAlign: "center",
    paddingTop: customVariables.contentPadding,
    fontSize: customVariables.fontSizeSmall
  }
});

type Props = Readonly<{
  image: ImageSourcePropType;
  title: string;
  subtitle?: string;
}>;

export const EmptyListComponent = (props: Props) => {
  return (
    <View style={styles.view}>
      <View spacer={true} />
      <Image source={props.image} />
      <Text style={styles.title}>{props.title}</Text>
      {props.subtitle && <Text style={styles.subtitle}>{props.subtitle}</Text>}
    </View>
  );
};
