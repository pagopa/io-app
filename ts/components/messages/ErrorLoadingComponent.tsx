import I18n from "i18n-js";
import { Text, View } from "native-base";
import * as React from "react";
import { Image, StyleSheet } from "react-native";
import customVariables from "../../theme/variables";

const styles = StyleSheet.create({
  view: {
    padding: customVariables.contentPadding,
    alignItems: "center"
  },
  title: {
    paddingTop: customVariables.contentPadding
  }
});

export const ErrorLoadingComponent = () => {
  return (
    <View style={styles.view}>
      <View spacer={true} />
      <Image
        source={require("../../../img/messages/empty-message-list-icon.png")}
      />
      <Text style={styles.title}>{I18n.t("messages.loadingErrorTitle")}</Text>
    </View>
  );
};
