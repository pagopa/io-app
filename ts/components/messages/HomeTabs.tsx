import { Tab, Tabs } from "native-base";
import React, { useState } from "react";
import { Animated, Platform, StyleSheet, View } from "react-native";
import { connect } from "react-redux";
import I18n from "../../i18n";
import { setMessagesArchivedState } from "../../store/actions/messages";
import { navigateToMessageRouterScreen } from "../../store/actions/navigation";
import { Dispatch } from "../../store/actions/types";
import { allMessagesSelector } from "../../store/reducers/entities/messages/allPaginated";
import { GlobalState } from "../../store/reducers/types";
import { makeFontStyleObject } from "../../theme/fonts";
import customVariables from "../../theme/variables";
import { isAndroid } from "../../utils/platform";

import { IOStyles } from "../core/variables/IOStyles";

import MessagesInbox from "./paginated/MessagesInbox";

const styles = StyleSheet.create({
  tabBarContainer: {
    elevation: 0,
    height: 40
  },
  tabBarUnderline: {
    borderBottomColor: customVariables.tabUnderlineColor,
    borderBottomWidth: customVariables.tabUnderlineHeight
  },
  tabBarUnderlineActive: {
    height: customVariables.tabUnderlineHeight,
    // borders do not overlap each other, but stack naturally
    marginBottom: -customVariables.tabUnderlineHeight,
    backgroundColor: customVariables.contentPrimaryBackground
  },
  searchDisableIcon: {
    color: customVariables.headerFontColor
  },
  activeTextStyle: {
    ...makeFontStyleObject(Platform.select, "600"),
    fontSize: isAndroid ? 16 : undefined,
    fontWeight: isAndroid ? "normal" : "bold",
    color: customVariables.brandPrimary
  },
  textStyle: {
    color: customVariables.brandDarkGray
  }
});

const AnimatedTabs = Animated.createAnimatedComponent(Tabs);

type OwnProps = {
  animatedTabScrollPositions: ReadonlyArray<Animated.Value>;
  onTabChange: (i: number) => void;
};

type Props = OwnProps &
  ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps>;

/**
 * Render Inbox, Deadlines and Archive tabs.
 */
const HomeTabs = ({
  animatedTabScrollPositions,
  messages,
  navigateToMessageDetail,
  onTabChange,
  updateMessagesArchivedState
}: Props) => {
  const [currentTab, setCurrentTab] = useState(0);
  // Update currentTab state when horizontal scroll is completed
  const handleOnChangeTab = (event: { i: number }) => {
    setCurrentTab(event.i);
    onTabChange(event.i);
  };

  // Disable longPress options the horizontal scroll overcome the 50% of the tab width
  const handleOnTabsScroll = (value: number) => {
    if (Math.abs(value - currentTab) > 0.5) {
      const nextTab = currentTab + (value - currentTab > 0 ? 1 : -1);
      setCurrentTab(nextTab);
      onTabChange(nextTab);
    }
  };

  return (
    <View style={IOStyles.flex}>
      <AnimatedTabs
        tabContainerStyle={[styles.tabBarContainer, styles.tabBarUnderline]}
        tabBarUnderlineStyle={styles.tabBarUnderlineActive}
        onScroll={handleOnTabsScroll}
        onChangeTab={handleOnChangeTab}
        initialPage={0}
      >
        <Tab
          activeTextStyle={styles.activeTextStyle}
          textStyle={styles.textStyle}
          heading={I18n.t("messages.tab.inbox")}
        >
          <MessagesInbox
            animated={{
              onScroll: Animated.event([
                {
                  nativeEvent: {
                    contentOffset: { y: animatedTabScrollPositions[0] }
                  }
                }
              ]),
              scrollEventThrottle: 8
            }}
            isActive={currentTab === 0}
            messages={messages}
            navigateToMessageDetail={navigateToMessageDetail}
            setMessagesArchivedState={updateMessagesArchivedState}
          />
        </Tab>
      </AnimatedTabs>
    </View>
  );
};

const mapStateToProps = (state: GlobalState) => ({
  messages: allMessagesSelector(state)
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  navigateToMessageDetail: (messageId: string) =>
    navigateToMessageRouterScreen({ messageId }),
  updateMessagesArchivedState: (
    ids: ReadonlyArray<string>,
    archived: boolean
  ) => dispatch(setMessagesArchivedState(ids, archived))
});

export default connect(mapStateToProps, mapDispatchToProps)(HomeTabs);
