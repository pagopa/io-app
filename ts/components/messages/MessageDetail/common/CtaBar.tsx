import * as O from "fp-ts/lib/Option";
import React from "react";
import { View, StyleSheet } from "react-native";
import { CommonServiceMetadata } from "../../../../../definitions/backend/CommonServiceMetadata";
import { ServiceId } from "../../../../../definitions/backend/ServiceId";
import { useIODispatch } from "../../../../store/hooks";
import {
  getPaymentExpirationInfo,
  PaymentData,
  UIMessageDetails,
  UIMessageId
} from "../../../../store/reducers/entities/messages/types";
import { UIService } from "../../../../store/reducers/entities/services/types";
import variables from "../../../../theme/variables";

import {
  getMessageCTA,
  isExpired,
  MessagePaymentExpirationInfo
} from "../../../../utils/messages";
import { HSpacer } from "../../../core/spacer/Spacer";
import { IOStyles } from "../../../core/variables/IOStyles";
import ExtractedCTABar from "../../../cta/ExtractedCTABar";
import CalendarEventButton from "./CalendarEventButton";
import PaymentButton from "./PaymentButton";

type Props = {
  isPaid: boolean;
  messageDetails: UIMessageDetails;
  service?: UIService;
  serviceMetadata?: CommonServiceMetadata;
};

const styles = StyleSheet.create({
  footerContainer: {
    overflow: "hidden",
    marginTop: -variables.footerShadowOffsetHeight,
    paddingTop: variables.footerShadowOffsetHeight
  }
});

// return a payment button only when the advice is not paid and the payment_data is defined
const renderPaymentButton = (
  paymentData?: PaymentData,
  messageId?: UIMessageId
) => {
  if (paymentData === undefined) {
    return null;
  }
  const { amount, noticeNumber, payee } = paymentData;
  return (
    <PaymentButton
      amount={amount}
      noticeNumber={noticeNumber}
      organizationFiscalCode={payee.fiscalCode}
      messageId={messageId}
    />
  );
};

function renderCalendarEventButton(
  isPaid: boolean,
  expirationInfo: MessagePaymentExpirationInfo,
  legacyMessage: UIMessageDetails["raw"],
  dueDate?: Date
) {
  if (dueDate === undefined || isPaid || isExpired(expirationInfo)) {
    return null;
  }
  return <CalendarEventButton message={legacyMessage} />;
}

/**
 * A smart component to show the action buttons of a message.
 * It requires context and full redux support.
 *
 * For messages with the proper configuration renders, on a row:
 * - a calendar icon
 * - a button to add/remove a calendar event
 * - a button to show/start a payment
 */
const CtaBar = ({
  isPaid,
  messageDetails,
  service,
  serviceMetadata
}: Props): React.ReactElement | null => {
  const dispatch = useIODispatch();
  // in case of medical prescription, we shouldn't render the CtaBar
  if (messageDetails.prescriptionData !== undefined) {
    return null;
  }
  const expirationInfo = getPaymentExpirationInfo(messageDetails);
  const { dueDate, markdown, paymentData, raw: legacyMessage } = messageDetails;

  const paymentButton = renderPaymentButton(paymentData, messageDetails.id);
  const calendarButton = renderCalendarEventButton(
    isPaid,
    expirationInfo,
    legacyMessage,
    dueDate
  );

  const footer1 = (paymentButton || calendarButton) && (
    // Added a wrapper to enable the usage of the component outside the Container of Native Base
    <View style={styles.footerContainer} pointerEvents={"box-none"}>
      <View style={[IOStyles.footer, IOStyles.row]}>
        {calendarButton}
        {paymentButton && calendarButton && <HSpacer size={16} />}
        {paymentButton}
      </View>
    </View>
  );
  const maybeCtas = getMessageCTA(
    markdown,
    serviceMetadata,
    service?.id as ServiceId
  );
  const footer2 = O.isSome(maybeCtas) && (
    // Added a wrapper to enable the usage of the component outside the Container of Native Base
    <View style={styles.footerContainer} pointerEvents={"box-none"}>
      <View testID={"CtaBar_withCTA"} style={[IOStyles.footer, IOStyles.row]}>
        <ExtractedCTABar
          ctas={maybeCtas.value}
          xsmall={false}
          dispatch={dispatch}
          serviceMetadata={serviceMetadata}
          service={service?.raw}
        />
      </View>
    </View>
  );
  return (
    <View testID={"CtaBar_withFooter"}>
      {footer2}
      {footer1}
    </View>
  );
};

export default CtaBar;
