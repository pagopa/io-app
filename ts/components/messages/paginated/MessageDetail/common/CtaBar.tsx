import { View as NBView } from "native-base";
import React from "react";
import { Platform, StyleSheet } from "react-native";
import DeviceInfo from "react-native-device-info";
import {
  getMessageCTA,
  isExpired,
  MessagePaymentExpirationInfo
} from "../../../../../utils/messages";
import {
  PaymentData,
  UIMessageDetails
} from "../../../../../store/reducers/entities/messages/types";
import { ServiceId } from "../../../../../../definitions/backend/ServiceId";
import { UIService } from "../../../../../store/reducers/entities/services/types";
import ExtractedCTABar from "../../../../cta/ExtractedCTABar";
import { useIODispatch } from "../../../../../store/hooks";
import PaymentButton from "../../../PaymentButton";
import CalendarEventButton from "../../../CalendarEventButton";
import { ServiceMetadata } from "../../../../../../definitions/backend/ServiceMetadata";

type Props = {
  expirationInfo: MessagePaymentExpirationInfo;
  isPaid: boolean;
  isPrescription: boolean;
  messageDetails: UIMessageDetails;
  service?: UIService;
  serviceMetadata?: ServiceMetadata;
};

const styles = StyleSheet.create({
  row: {
    flexDirection: "row",
    paddingBottom: Platform.OS === "ios" && DeviceInfo.hasNotch() ? 28 : 15
  }
});

// return a payment button only when the advice is not paid and the payment_data is defined
const renderPaymentButton = (paymentData?: PaymentData) => {
  if (paymentData === undefined) {
    return null;
  }
  const { amount, noticeNumber, payee } = paymentData;
  return (
    <PaymentButton
      amount={amount}
      noticeNumber={noticeNumber}
      organizationFiscalCode={payee.fiscalCode}
    />
  );
};

function renderCalendarEventButton(
  isPaid: boolean,
  expirationInfo: MessagePaymentExpirationInfo,
  legacyMessage: UIMessageDetails["raw"],
  dueDate?: Date
) {
  if (dueDate === undefined || isPaid || isExpired(expirationInfo)) {
    return null;
  }
  return <CalendarEventButton message={legacyMessage} />;
}

/**
 * A smart component to show the action buttons of a message.
 * It requires context and full redux support.
 *
 * For messages with the proper configuration renders, on a row:
 * - a calendar icon
 * - a button to add/remove a calendar event
 * - a button to show/start a payment
 */
const CtaBar = ({
  isPaid,
  expirationInfo,
  isPrescription,
  messageDetails,
  service,
  serviceMetadata
}: Props): React.ReactElement | null => {
  const dispatch = useIODispatch();
  // in case of medical prescription, we shouldn't render the CtaBar
  if (isPrescription) {
    return null;
  }

  const { dueDate, markdown, paymentData, raw: legacyMessage } = messageDetails;

  const paymentButton = renderPaymentButton(paymentData);
  const calendarButton = renderCalendarEventButton(
    isPaid,
    expirationInfo,
    legacyMessage,
    dueDate
  );

  const footer1 = (paymentButton || calendarButton) && (
    <NBView footer={true} style={styles.row}>
      {calendarButton}
      {paymentButton && calendarButton && <NBView hspacer={true} />}
      {paymentButton}
    </NBView>
  );
  const maybeCtas = getMessageCTA(
    markdown,
    serviceMetadata,
    service?.id as ServiceId
  );
  const footer2 = maybeCtas.isSome() && (
    <NBView testID={"CtaBar_withCTA"} footer={true} style={styles.row}>
      <ExtractedCTABar
        ctas={maybeCtas.value}
        xsmall={false}
        dispatch={dispatch}
        serviceMetadata={serviceMetadata}
        service={service?.raw}
      />
    </NBView>
  );
  return (
    <NBView testID={"CtaBar_withFooter"}>
      {footer2}
      {footer1}
    </NBView>
  );
};

export default CtaBar;
