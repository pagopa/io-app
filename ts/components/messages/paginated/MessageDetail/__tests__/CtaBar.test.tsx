import React from "react";
import configureMockStore from "redux-mock-store";
import { NavigationParams } from "react-navigation";

import { toUIMessageDetails } from "../../../../../store/reducers/entities/messages/transformers";
import { paymentValidInvalidAfterDueDate } from "../../../../../__mocks__/message";
import { toUIService } from "../../../../../store/reducers/entities/services/transformers";
import { service_1 } from "../../../../../__mocks__/messages";
import { MessagePaymentExpirationInfo } from "../../../../../utils/messages";
import { appReducer } from "../../../../../store/reducers";
import { applicationChangeState } from "../../../../../store/actions/application";
import { GlobalState } from "../../../../../store/reducers/types";
import { renderScreenFakeNavRedux } from "../../../../../utils/testWrapper";
import ROUTES from "../../../../../navigation/routes";
import I18n from "../../../../../i18n";

import CtaBar from "../common/CtaBar";

jest.useFakeTimers();

const uiMessageDetails = toUIMessageDetails(paymentValidInvalidAfterDueDate);
const uiService = toUIService(service_1);

const defaultProps = {
  expirationInfo: {
    kind: "EXPIRABLE",
    expireStatus: "VALID",
    dueDate: uiMessageDetails.dueDate
  } as MessagePaymentExpirationInfo,
  isPaid: true,
  messageDetails: uiMessageDetails,
  service: uiService,
  servicesMetadata: undefined,
  isPrescription: false
};

describe("the `CtaBar` component", () => {
  describe("when neither `paymentData` nor `dueDate` are defined for the message details", () => {
    it("should match the snapshot (no buttons)", () => {
      const noPaymentMessage = {
        ...uiMessageDetails,
        paymentData: undefined,
        dueDate: undefined
      };
      const { component } = renderComponent({
        ...defaultProps,
        messageDetails: noPaymentMessage
      });
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe("when `paymentData` is defined", () => {
    describe("and `isPaid` is true", () => {
      it("should render the payment button", () => {
        const { component } = renderComponent(defaultProps);
        expect(
          component.queryByText(I18n.t("messages.cta.seeNotice"))
        ).not.toBeNull();
      });
    });
    describe("and `isPaid` is false", () => {
      it("should render the payment button", () => {
        const { component } = renderComponent({
          ...defaultProps,
          isPaid: false
        });
        expect(
          component.queryByText(I18n.t("messages.cta.seeNotice"))
        ).not.toBeNull();
      });
    });
  });

  describe("when `dueDate` is defined", () => {
    describe("and `isPaid` is true", () => {
      it("should not render the calendar button", () => {
        const { component } = renderComponent(defaultProps);
        expect(
          component.queryByText(I18n.t("messages.cta.reminder"))
        ).toBeNull();
      });
    });
    describe("and `isPaid` is false", () => {
      it("should render the calendar button", () => {
        const { component } = renderComponent({
          ...defaultProps,
          isPaid: false
        });
        expect(
          component.queryByText(I18n.t("messages.cta.reminder"))
        ).not.toBeNull();
      });
    });
    describe("and the payment has expired", () => {
      it("should not render the calendar button", () => {
        const { component } = renderComponent({
          ...defaultProps,
          expirationInfo: {
            ...defaultProps.expirationInfo,
            expireStatus: "EXPIRED"
          }
        });
        expect(
          component.queryByText(I18n.t("messages.cta.reminder"))
        ).toBeNull();
      });
    });
  });

  describe("when at least one CTA is defined in the markdown", () => {
    const markdown = `---
en:
    cta_1: 
        text: "DON'T PANIC!"
        action: "ioit://PROFILE_MAIN"
---
                      ` as any;

    it("should render it", () => {
      const { component } = renderComponent({
        ...defaultProps,
        messageDetails: {
          ...defaultProps.messageDetails,
          markdown
        }
      });
      expect(component.queryByText("DON'T PANIC!")).not.toBeNull();
    });
  });
});

const renderComponent = (props: React.ComponentProps<typeof CtaBar>) => {
  const globalState = appReducer(undefined, applicationChangeState("active"));
  const mockStore = configureMockStore<GlobalState>();
  const store: ReturnType<typeof mockStore> = mockStore(globalState);

  return {
    component: renderScreenFakeNavRedux<GlobalState, NavigationParams>(
      () => <CtaBar {...props} />,
      ROUTES.MESSAGE_DETAIL,
      {},
      store
    ),
    store
  };
};
