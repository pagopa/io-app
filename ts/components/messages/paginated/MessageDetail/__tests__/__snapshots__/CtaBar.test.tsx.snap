// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the \`CtaBar\` component when neither \`paymentData\` nor \`dueDate\` are defined for the message details should match the snapshot (no buttons) 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      Object {
        "backgroundColor": "#eee",
        "flex": 1,
        "flexDirection": "column-reverse",
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        forwardedRef={[Function]}
        pointerEvents="none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": Array [
              Object {
                "translateX": 1000000,
              },
              Object {
                "translateY": 1000000,
              },
            ],
          }
        }
      >
        <View
          accessibilityElementsHidden={true}
          style={
            Object {
              "backgroundColor": "#fff",
              "flex": 1,
            }
          }
        />
      </View>
      <View
        forwardedRef={[Function]}
        pointerEvents="box-only"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "transform": Array [
              Object {
                "translateX": 0,
              },
              Object {
                "translateY": 0,
              },
            ],
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          style={
            Object {
              "backgroundColor": "#fff",
              "flex": 1,
            }
          }
        >
          <View
            style={Object {}}
            testID="CtaBar_withFooter"
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
